### ISSUE TITLE
Investigate Compiler Performance

### ISSUE THREAD

**Issue by mikearnaldi** [Effect-TS/core#550]
The current HKT encoding seems to be very heavy on the compiler, investigate if it is possible to improve, eg:

```ts
import { pipe, tuple } from "@effect-ts/core/Function"
import * as NA from "@effect-ts/core/NonEmptyArray"
import * as R from "@effect-ts/core/Record"

type License = "CC" | "MIT"

interface Hero {
  name: string
  license: License
}

function hero(name: string, license: License): Hero {
  return { name, license }
}

const toRecord = R.fromFoldable(NA.getAssociative<Hero>(), NA.Foldable)

const grouped = pipe(
  [hero("Johannes", "CC"), hero("Mike", "MIT")],
  NA.map((h) => tuple(h.license, NA.single(h))),
  toRecord
)
```

Takes about 10 seconds for VSCode LSP to keep up with changes

**Comment by mikearnaldi** [comment]
Conditional types and especially tuple recursion are the problem, the only practical solution is to remove support for transformers #552

**Comment by mikearnaldi** [comment]
closed with #556
